src/main/java/cloud/eppo/ConfigurationWrangler.java

```diff
@@ -0,0 +1,241 @@
+package cloud.eppo;
+
+import cloud.eppo.api.Configuration;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConfigurationWrangler {
+  private static final Logger log = LoggerFactory.getLogger(ConfigurationWrangler.class);
+
+  private final ConfigurationRequestor requestor;
+  private final IConfigurationStore configurationStore;
+
+  // This reference is **the** reference to the current configuration.
+  private volatile Configuration config;
+
+  ConfigurationWrangler(
+      ConfigurationRequestor requestor,
+      IConfigurationStore persistentSource,
+      Configuration initialConfiguration) {
+    this.requestor = requestor;
+    this.configurationStore = persistentSource;
+
+    setConfig(initialConfiguration != null ? initialConfiguration : Configuration.emptyConfig());
+  }
+
+  private final ReentrantLock lock = new ReentrantLock();
+
+  private void setConfig(Configuration configuration) {
+    lock.lock();
+    try {
+      config = configuration;
+
+      // Fire-and-forget save call to the config store. It may or may not block (non-blocking is
+      // best)
+      configurationStore.saveConfiguration(config);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Loads the configuration from the `ConfigurationRequestor` and saves to "persistent" storage
+   * using `IConfigurationStore`
+   *
+   * <p>This method attempts to load from the `IConfigurationStore` if it was unable to load from
+   * the API.
+   */
+  void load() {
+    try {
+      Configuration newConfig = requestor.load(config);
+      setConfig(newConfig);
+    } catch (RuntimeException e) {
+      log.error("Failed to load configuration", e);
+
+      // Try getting config from the cache. It loads async, so we'll use a blocking queue.
+      BlockingQueue<Configuration> bq = new LinkedBlockingQueue<>(1);
+
+      configurationStore.load(
+          new IConfigurationStore.CacheLoadCallback() {
+            @Override
+            public void onSuccess(Configuration result) {
+              bq.add(result);
+            }
+
+            @Override
+            public void onFailure(String errorMessage) {
+              throw new RuntimeException(errorMessage);
+            }
+          });
+
+      try {
+        // Only set the config reference as this value came from the cache
+        config = bq.take();
```

Comment:
Oh cool! So this will wait until something has been added to the queue? Nifty.

I wonder if we should have timeouts for this stuff, settable in Configuration but with some reasonable defaults.

Reply:
That's right. This seemed like the cleanest way to turn the callbacks into sync. Java `Future`s have a lot of this built-in, but that would require the SDK be bumped up to 24 (from 21) in the Android SDK :( There is a way to only wait a specified amount of time before giving up. We'll have to add another parameter to the `BaseEppoClient` constructor to plumb through the timeout, but that's cleaned up by the `Builder`s in the java-server- and android SDKs